### ISSUE TITLE
Gateway schema refresh issue with TypeGraphQL

### ISSUE THREAD

**Issue by franck-romano** [mercurius-js/mercurius#379]
Hey there,

First thank you for the good work on this awesome library 🙏 , the fastify eco-system is just neat !

----

I have an issue that's making me crazy and I'm seeking your help on this.

I have a gateway running in federation mode like this:

```
server.register(mercurius, {
    jit: 1,
    graphiql: isProduction ? false : 'playground',
    gateway: {
      services: [
        {
          name: 'my-service',
          url: 'http://localhost:8080/graphql',
          rewriteHeaders: (headers: IncomingHttpHeaders) => {
            if (headers.authorization) {
              return {
                authorization: headers.authorization
              };
            }
            return {};
          },
          mandatory: true
        }
      ],
      pollingInterval: POOLING_INTERVAL_IN_MS
    }
  });
```

And a downstream service using mercurius (not in gateway mode) with schema generated through [type-graphql](https://github.com/MichalLytek/type-graphql).

The generated schema is:

```
# -----------------------------------------------
# !!! THIS FILE WAS GENERATED BY TYPE-GRAPHQL !!!
# !!!   DO NOT MODIFY THIS FILE BY YOURSELF   !!!
# -----------------------------------------------

directive @extends on INTERFACE | OBJECT

directive @external on FIELD_DEFINITION | OBJECT

directive @key(fields: String!) on INTERFACE | OBJECT

directive @provides(fields: String!) on FIELD_DEFINITION

directive @requires(fields: String!) on FIELD_DEFINITION

type User {
  email: String!
  id: ID!
  name: String!
}

type Query {
  findUserByEmail(email: String!): User!
}

```

**Issue**:
1. The gateway and the downstream service are running
2. I add a new mutation/query whatever to the downstream service
3. I restart the downstream service to regenerate the schema
4. And when I try to query the gateway with the newly query/mutation or even with the existing one, I got an error saying "**Cannot return null for non-nullable field Query.myNewQuery"**

The only way to make it works is to restart the gateway (or triggering a new deployment) which is not very convenient 😞 .

I'm suspecting a schema issue since the schema is not regenerated even on `playground`.

Do you have any inputs ?

Thanks for your time ! 😄 

Franck

**Comment by mcollina** [comment]
Thanks for reporting this. I do not know what type-graphql does differently and what is causing this.... we have tests in place to verify this behavior, so it _should_ work. You are probably in the best position to diagnose what's the root cause of this and send a PR to fix it. Would you like to give it a go?

**Comment by franck-romano** [comment]
Hello !

> Thanks for reporting this. I do not know what type-graphql does differently and what is causing this.... we have tests in place to verify this behavior, so it should work.

You're welcome.

> You are probably in the best position to diagnose what's the root cause of this and send a PR to fix it. Would you like to give it a go?

Yup I can give it a try, however I need some time to setup a dummy repository so I can iterate on it.
I'll keep you up-to-date :)

:wave:

**Comment by franck-romano** [comment]
I found something very strange: the schema is properly refreshed once something is added to the downstream service but I keep on having this error: "Cannot return null for non-nullable field Query...".

I noticed that the query/mutation  never reaches the downstream service.

Here is the hypothesis that I will try to follow: 

- There might be a difference between the gateway's boostrap behaviour and the `refresh` method.

I will dig into this direction but if you have some thoughs about this - let me know, I might be wrong 😅 .

:wave:

**Comment by mcollina** [comment]
Seems promising, keep going!

**Comment by franck-romano** [comment]
I managed to find a little difference:

As we can see [here](https://github.com/mercurius-js/mercurius/blob/master/lib/gateway.js#L307) `Query`, `Mutation` and `Subscription` fields are set to `null` before defining the resolvers.

However, if we have a look at the [`refresh()` method](https://github.com/mercurius-js/mercurius/blob/master/lib/gateway.js#L371), those fields are not set to  `null`.

Hence, those field values  are set to the name of the downstream service.

I applied the same pattern in my local mercurius dependency (i.e set to `null` mentioned fields) and it solved the refresh issue meaning that I can add mutations, queries and types without re-starting the gateway.

I added this in the `refresh()` method just before defining the resolvers:

```   
typeToServiceMap.Query = null;
typeToServiceMap.Mutation = null;
typeToServiceMap.Subscription = null;

defineResolvers(schema, typeToServiceMap, serviceMap, typeFieldsToService);
```

(I didn't run the mercurius test suit yet since I'm a bit more in an exploration phase than on solving a precise problem but that should be done once we found the root cause of this issue).

I'm not able to measure the impacts this could have so I would like to get your thoughts on it 😄 .

WDYT ?
